platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheGracefulDegradationIntegrationTest.groovy

```diff
@@ -0,0 +1,290 @@
+                   println("Project path is \${project.path}")
+               }
+            }
+        """
+
+        when:
+        configurationCacheRun "a"
+
+        then:
+        configurationCache.assertNoConfigurationCache()
+        problems.assertResultHtmlReportHasProblems(result) {
+            totalProblemsCount = 1
+            withProblem("Build file 'build.gradle': line 5: invocation of 'Task.project' at execution time is unsupported.")
+            withIncompatibleTask(":a", "Project access.")
+        }
+
+        and:
+        outputContains("Project path is :")
+        assertConfigurationCacheDegradation("task `:a` of type `DegradingTask`")
+    }
+
+    def "a task can require CC degradation for multiple reasons"() {
+        def configurationCache = newConfigurationCacheFixture()
+        buildFile """
+            ${taskWithInjectedDegradationController()}
+            tasks.register("a", DegradingTask) { task ->
+                def shouldAccessTaskProjectProvider = providers.systemProperty("accessTaskProject").map { Boolean.parseBoolean(it) }.orElse(false)
+                def shouldAccessTaskDependenciesProvider = providers.systemProperty("accessTaskDependencies").map { Boolean.parseBoolean(it) }.orElse(false)
+
+                getDegradationController().requireConfigurationCacheDegradation(task, shouldAccessTaskProjectProvider.map { it ? "Project access" : null })
+                getDegradationController().requireConfigurationCacheDegradation(task, shouldAccessTaskDependenciesProvider.map { it ? "TaskDependencies access" : null })
+
+                doLast {
+                    if (shouldAccessTaskProjectProvider.get()) {
+                        it.project
+                        println "Task's project accessed!"
+                    }
+                    if (shouldAccessTaskDependenciesProvider.get()) {
+                        it.taskDependencies
+                        println "Task's dependencies accessed!"
+                    }
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun("a", *args)
+
+        then:
+        configurationCache.assertNoConfigurationCache()
+        problems.assertResultHtmlReportHasProblems(result) {
+            totalProblemsCount = expectedProblems.size()
+            expectedProblems.forEach { withProblem(it) }
+            withIncompatibleTask(":a", degradationReason)
+        }
+
+        where:
+        args                                                          | expectedOutputs                                               | expectedProblems                                                                                                | degradationReason
+        ["-DaccessTaskProject=true"]                                  | ["Task's project accessed!"]                                  | ["Build file 'build.gradle': line 10: invocation of 'Task.project' at execution time is unsupported."]          | "Project access."
```

Comment:
💅  The `expectedProblems` column makes this table rather large

There are many ways to structure the `where:` clause, maybe some of them could offer a different structure
https://spockframework.org/spock/docs/2.3/data_driven_testing.html

Alternatively, consider replacing `expectedProblems` column with a boolean condition and use it do decide whether to expect an extra problem already inside the `assertResultHtmlReportHasProblems {}` lambda.